lib/facter/resolvers/aio_agent_version.rb

```diff
@@ -3,6 +3,7 @@
 module Facter
   module Resolvers
     class AioAgentVersion < BaseResolver
+      @semaphore = Mutex.new
```

Comment:
could we have the semaphore in a base class?

Reply:
We need to have concurrency between resolvers, but no concurrency inside a resolver, thus we need a semaphore for each resolver. 

The tricky part comes when we need to have static methods (because we don't want to create a new resolver in each fact). The semaphore need to be accessible in the `BaseResolver`, but cannot be instantiated in it as class variables (@semaphore) and `initialise` methods are never called when calling static methods from "BaseResolver" children. 

Here is a small example
```ruby
class Base
  def initialize
    puts "baser init"
    @semaphore = Mutex.new
  end

  class << self
    def work
      puts @semaphore.inspect
      do_work
    end
  end
end


class A < Base
  def initialize
    puts "A init"
  end

  class << self
    def do_work
      puts "work in A"
    end
  end
end

class B < Base
  def intialize
    puts "B init"
  end

  class << self
    def do_work
      puts "work in B"
    end
  end
end

A.work
B.work
```

results
```
nil
work in A
nil
work in B
```

An option would be to create a mechanism similar to the one used for [logger](https://github.com/puppetlabs/facter/blob/1f90f3f4fad599eb4f4dadbb61aca18381078de6/lib/facter/resolvers/base_resolver.rb#L6) in `BaseResolver`, but this would require an extra method call in each resolver.

Reply:
I think that could be achieved with modules and some Ruby magic:

https://gist.github.com/gimmyxd/d0685c7bdc2fe2c6f92bcf55e953520e